File: hackathon/kakaopay/submission/src/common/trade-plan.mjs
/**
 * 매매 계획 계산기 (순수 함수, 네트워크 없음).
 *
 * "매입가·매도가 추천"의 정직한 대응물 — 가격 예측이 아니라 **규칙**이다:
 * - 진입: 3분할 매수 밴드 (1차 30% 현재가, 2차 30% -5%, 3차 40% -10%)
 * - 손절: 평균 매수가 대비 -7% (전량)
 * - 익절: 평균 매수가 대비 +20% 도달 시 절반 매도, 잔여분은 손절선을 본전으로 상향
 * 모든 가격은 KRX 호가단위로 내림 처리한다.
 */

/** KRX 호가단위 (2023-01 개편 기준, 코스피·코스닥 공통) */
export function krxTickSize(price) {
  if (price < 2_000) return 1;
  if (price < 5_000) return 5;
  if (price < 20_000) return 10;
  if (price < 50_000) return 50;
  if (price < 200_000) return 100;
  if (price < 500_000) return 500;
  return 1_000;
}

export function roundToTick(price) {
  const tick = krxTickSize(price);
  return Math.floor(price / tick) * tick;
}

const ENTRY_BANDS = [
  { label: '1차', budgetShare: 0.3, priceRatio: 1.0 },
  { label: '2차', budgetShare: 0.3, priceRatio: 0.95 },
  { label: '3차', budgetShare: 0.4, priceRatio: 0.9 },
];
const STOP_LOSS_RATIO = 0.93;   // 평균 매수가 -7%
const TAKE_PROFIT_RATIO = 1.2;  // 평균 매수가 +20%

/**
 * @param {{ price: number, budget?: number }} input 현재가(원), 총 예산(원 — 선택)
 * @returns 진입 밴드·손절/익절 기준가(·예산 시 수량). 예산이 1주에도 못 미치면 { feasible: false }.
 *   budget 생략 시: 수량 없이 주당 가격 레벨만 반환 (평균가는 30/30/40 가중 가정).
 */
export function computeTradePlan({ price, budget }) {
  if (!Number.isFinite(price) || price <= 0) throw new Error(`price 가 유효하지 않습니다: ${price}`);
  const hasBudget = budget !== undefined && budget !== null;
  if (hasBudget && (!Number.isFinite(budget) || budget <= 0)) throw new Error(`budget 이 유효하지 않습니다: ${budget}`);

  if (!hasBudget) {
    // 가격 레벨 전용 모드 — 평균가는 밴드 전량 체결 가정(30/30/40 가중)
    const entries = ENTRY_BANDS.map(({ label, budgetShare, priceRatio }) => ({
      label,
      targetPrice: roundToTick(price * priceRatio),
      budgetShare,
    }));
    const avgEntry = entries.reduce((a, e) => a + e.targetPrice * e.budgetShare, 0);
    return {
      feasible: true,
      mode: 'price-levels-only',
      entries,
      avgEntryIfAllFilled: Math.round(avgEntry),
      exits: {
        stopLoss: {
          price: roundToTick(avgEntry * STOP_LOSS_RATIO),
          rule: '평균 매수가 -7% 도달 시 전량 매도 — 예외를 만들지 않는다',
        },
        takeProfitFirst: {
          price: roundToTick(avgEntry * TAKE_PROFIT_RATIO),
          rule: '평균 매수가 +20% 도달 시 절반 매도, 잔여분 손절선은 본전(평균 매수가)으로 상향',
        },
      },
      notes: [
        '이 가격들은 예측이 아니라 규칙이다 — 2·3차 밴드에 안 오면 그 비중은 집행하지 않는 것이 원칙',
        '수량은 예산을 주면 계산된다 (budget 인자). 모든 가격은 KRX 호가단위 내림, 수수료·세금 미반영',
      ],
    };
  }

  const entries = ENTRY_BANDS.map(({ label, budgetShare, priceRatio }) => {
    const targetPrice = roundToTick(price * priceRatio);
    const alloc = budget * budgetShare;
    const quantity = Math.floor(alloc / targetPrice);
    return { label, targetPrice, quantity, amount: quantity * targetPrice };
  });

  const totalQuantity = entries.reduce((a, e) => a + e.quantity, 0);
  const totalAmount = entries.reduce((a, e) => a + e.amount, 0);
  if (totalQuantity === 0) {
    return {
      feasible: false,
      reason: `예산 ${budget.toLocaleString()}원으로는 1주(현재가 ${price.toLocaleString()}원)도 매수할 수 없습니다`,
    };
  }
  const avgEntry = totalAmount / totalQuantity;

  return {
    feasible: true,
    entries,
    totalQuantity,
    totalAmount: Math.round(totalAmount),
    avgEntryIfAllFilled: Math.round(avgEntry),
    exits: {
      stopLoss: {
        price: roundToTick(avgEntry * STOP_LOSS_RATIO),
        rule: '평균 매수가 -7% 도달 시 전량 매도 — 예외를 만들지 않는다',
      },
      takeProfitFirst: {
        price: roundToTick(avgEntry * TAKE_PROFIT_RATIO),
        rule: '평균 매수가 +20% 도달 시 절반 매도, 잔여분 손절선은 본전(평균 매수가)으로 상향',
      },
    },
    notes: [
      '이 가격들은 예측이 아니라 규칙이다 — 2·3차 밴드에 안 오면 그 비중은 집행하지 않는 것이 원칙',
      '모든 가격은 KRX 호가단위 내림 적용, 수수료·세금 미반영',
    ],
  };
}

File: screening-backtest-service/src/screening_backtest/core/trade_plan.py
"""Pure trade-plan domain functions.

Reimplemented faithfully from the investment-service Java domain
(`TradePlanPolicy.java`): 3 tranche entries at 100%/95%/90% of current price
with weights 30%/30%/40%, KRX tick rounding (round DOWN), stop-loss =
avg entry * 0.93, take-profit = avg entry * 1.20. Fees/taxes not modelled.

All prices are Korean won (integer). Functions are import-testable and pure.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from decimal import ROUND_FLOOR, ROUND_HALF_UP, Decimal
from typing import Optional

# KRX tick bands (2023-01 reform, KOSPI/KOSDAQ common). Ordered ascending.
# (upper_exclusive, tick). A price below `upper` uses `tick`.
_TICK_BANDS: list[tuple[Decimal, Decimal]] = [
    (Decimal(2_000), Decimal(1)),
    (Decimal(5_000), Decimal(5)),
    (Decimal(20_000), Decimal(10)),
    (Decimal(50_000), Decimal(50)),
    (Decimal(200_000), Decimal(100)),
    (Decimal(500_000), Decimal(500)),
]
_TOP_TICK = Decimal(1_000)

# Entry bands: (label, budget_share, price_ratio)
_ENTRY_BANDS: list[tuple[str, Decimal, Decimal]] = [
    ("1차", Decimal("0.30"), Decimal("1")),
    ("2차", Decimal("0.30"), Decimal("0.95")),
    ("3차", Decimal("0.40"), Decimal("0.90")),
]

STOP_LOSS_RATIO = Decimal("0.93")
TAKE_PROFIT_RATIO = Decimal("1.20")


def krx_tick_size(price: Decimal) -> Decimal:
    """Return the KRX tick size for a price (2023-01 reform table)."""
    for upper, tick in _TICK_BANDS:
        if price < upper:
            return tick
    return _TOP_TICK


def krx_tick_round_down(price) -> int:
    """Round a price DOWN to the nearest KRX tick. Returns an int (won).

    Examples: 797000 -> tick 1000 -> 797000; 49999 -> tick 50 -> 49950;
    50000 -> tick 100 -> 50000.
    """
    p = Decimal(str(price))
    tick = krx_tick_size(p)
    floored = (p / tick).to_integral_value(rounding=ROUND_FLOOR) * tick
    return int(floored.to_integral_value(rounding=ROUND_FLOOR))


@dataclass
class EntryBand:
    label: str
    price: int
    budget_share: float
    quantity: Optional[int] = None
    amount: Optional[int] = None

    def to_dict(self) -> dict:
        return {
            "label": self.label,
            "price": self.price,
            "budgetShare": self.budget_share,
            "quantity": self.quantity,
            "amount": self.amount,
        }


@dataclass
class TradePlan:
    feasible: bool
    entries: list[EntryBand] = field(default_factory=list)
    total_quantity: Optional[int] = None
    total_amount: Optional[int] = None
    avg_entry: int = 0
    stop_loss: int = 0
    take_profit: int = 0
    reason: Optional[str] = None

    def to_dict(self) -> dict:
        return {
            "feasible": self.feasible,
            "reason": self.reason,
            "entries": [e.to_dict() for e in self.entries],
            "totalQuantity": self.total_quantity,
            "totalAmount": self.total_amount,
            "avgEntry": self.avg_entry,
            "stopLoss": self.stop_loss,
            "takeProfit": self.take_profit,
        }


def _half_up_int(value: Decimal) -> int:
    return int(value.to_integral_value(rounding=ROUND_HALF_UP))


def trade_plan(current_price, budget=None) -> TradePlan:
    """Compute the 3-tranche trade plan.

    If ``budget`` is None, returns a price-levels-only plan whose avg entry is
    the weighted (30/30/40) average of the tick-rounded band prices.
    Otherwise allocates the budget across bands, floors share quantities, and
    derives avg entry from filled amounts.
    """
    current = Decimal(str(current_price))
    if current <= 0:
        raise ValueError(f"currentPrice must be positive: {current_price}")

    if budget is None:
        return _price_levels_only(current)

    budget_d = Decimal(str(budget))
    if budget_d <= 0:
        raise ValueError(f"budget must be positive: {budget}")

    entries: list[EntryBand] = []
    total_quantity = 0
    total_amount = 0
    for label, share, ratio in _ENTRY_BANDS:
        target = krx_tick_round_down(current * ratio)
        qty = int((budget_d * share) // target) if target > 0 else 0
        amount = target * qty
        entries.append(EntryBand(label, target, float(share), qty, amount))
        total_quantity += qty
        total_amount += amount

    if total_quantity == 0:
        return TradePlan(
            feasible=False,
            entries=entries,
            reason=(
                f"budget {budget_d} cannot buy even 1 share "
                f"(current {current})"
            ),
        )

    avg_entry_d = Decimal(total_amount) / Decimal(total_quantity)
    return TradePlan(
        feasible=True,
        entries=entries,
        total_quantity=total_quantity,
        total_amount=total_amount,
        avg_entry=_half_up_int(avg_entry_d),
        stop_loss=krx_tick_round_down(avg_entry_d * STOP_LOSS_RATIO),
        take_profit=krx_tick_round_down(avg_entry_d * TAKE_PROFIT_RATIO),
    )


def _price_levels_only(current: Decimal) -> TradePlan:
    entries: list[EntryBand] = []
    avg_entry_d = Decimal(0)
    for label, share, ratio in _ENTRY_BANDS:
        target = krx_tick_round_down(current * ratio)
        entries.append(EntryBand(label, target, float(share)))
        avg_entry_d += Decimal(target) * share
    return TradePlan(
        feasible=True,
        entries=entries,
        avg_entry=_half_up_int(avg_entry_d),
        stop_loss=krx_tick_round_down(avg_entry_d * STOP_LOSS_RATIO),
        take_profit=krx_tick_round_down(avg_entry_d * TAKE_PROFIT_RATIO),
    )
